File: leetcode234/leetcode234.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        tmp = None
        cur = head
        prev = None

        while cur != None:
            tmp = cur.next
            cur.next = prev
            prev = cur
            cur = tmp
        
        return prev
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head
        pre = head
        while fast and fast.next:
            fast = fast.next.next
            pre = slow
            slow = slow.next
        
        pre.next = None

        cur1 = head
        cur2 = self.reverseList(slow)

        while cur1 and cur2:
            if cur1.val != cur2.val:
                return False
            cur1 = cur1.next
            cur2 = cur2.next
        return True

File: leetcode234/leetcode234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:

    ListNode* reverseList(ListNode* head) {
        ListNode* cur = head;
        ListNode* prev = nullptr;
        ListNode* tmp = nullptr;
        while (cur)
        {
            tmp = cur -> next;
            cur -> next = prev;
            prev = cur;
            cur = tmp;
        }
        return prev;
    }
    bool isPalindrome(ListNode* head) {
        if (head == nullptr || head -> next == nullptr) return true;
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* pre = head;
        while(fast && fast -> next){
            fast = fast -> next -> next;
            pre = slow;
            slow = slow -> next;
        }

        pre -> next = nullptr;

        ListNode* head1 = head;
        ListNode* head2 = reverseList(slow);
        ListNode* cur1 = head1;
        ListNode* cur2 = head2;
        while(cur1 && cur2){
            if (cur1 -> val != cur2 -> val) return false;
            cur1 = cur1 -> next;
            cur2 = cur2 -> next;
        }
        return true;
    }
};